### PR TITLE
feat: profiling v2 [MD-27]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

Profiling V2 ([Project Doc](https://github.com/azhou-determined/notes/blob/master/profiling-v2/profiling-v2.md))

Individual commits should have already been reviewed, this is the final feature branch to main PR.

Major changes:

- Functionality for timing metrics in the Determined profiler was dropped. Determined profiler is now only responsible for system metrics.
- Determined Profiler now lives in Core API. 
- Determined profiler's system metrics now use backend's generic metrics framework

## Test Plan

This feature should be tested across a few different Trial APIs

### PyTorch

For `mnist_pytorch`:
- Change `determined/examples/tutorials/mnist_pytorch/train.py` to add `profiling_enabled`:
```
trainer.fit(max_length=max_length, latest_checkpoint=latest_checkpoint, profiling_enabled=True)
```
- Submit the experiment. Go to the "Profiler" tab for that experiment in the Web UI. Verify that "System Metrics" is rendered with metric values.

### TF Keras
- In `determined/examples/computer_vision/iris_tf_keras/distributed.yaml`, enabled profiling:
```
profiling:
  enabled: true
```
- Submit the experiment. Go to the "Profiler" tab for that experiment in the Web UI. Verify that "System Metrics" is rendered with metric values.
image

### Core API
- Submit a Core API experiment:

```python
import time
import logging

import determined as det
from determined import core


def main(core_context):
    core_context.profiler.on(sampling_interval=0.1, samples_per_report=10)
    for batch in range(30):
        steps_completed = batch + 1
        if steps_completed % 5 == 0:
            core_context.train.report_training_metrics(
                steps_completed=steps_completed, metrics={"x": batch}
            )
        if steps_completed % 10 == 0:
            core_context.train.report_validation_metrics(steps_completed=steps_completed, metrics={"x": batch})
        time.sleep(1)
    core_context.profiler.off()


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format=det.LOG_FORMAT)
    with core.init() as core_context:
        main(core_context=core_context)
```
- Submit the experiment. Go to the "Profiler" tab for that experiment in the Web UI. Verify that "System Metrics" is rendered with metric values.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by gt2345 — APPROVED** [review]
WebUI stamp

File: docs/model-dev-guide/profiling.rst
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:19]:
```suggestion
A training job can be instrumented at many different layers, from generic system-level
```
or
```suggestion
A training job can be profiled at many different layers, from generic system-level
```

File: docs/model-dev-guide/profiling.rst
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:56]: I'm vaguely uncomfortable with "current", and I can't pinpoint the concept I think it's meant to indicate. I believe that the profiler aggregates average utilization over each interval. But I think that not specifying the exact method and letting users read some docstrings or code if they really care is a fine choice.

So I guess my recommendation is to just drop the word "current" wherever it appears.

Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:56]: I like the inclusion of units, BTW.

File: docs/model-dev-guide/profiling.rst
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:58]: I don't know what "change rate" means. Here, units are good enough that I'd probably understand if it just said "bytes sent syntemwide". The [wikipedia page](https://en.wikipedia.org/wiki/Data-rate_units) about units doesn't give me any words I like.

File: docs/model-dev-guide/profiling.rst
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:75]: This could be a place to reference how this can be configured.

File: docs/model-dev-guide/profiling.rst
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:35]: Is it appropriate to provide references here for how profiler metrics can be inspected?

File: docs/model-dev-guide/profiling.rst
Comment (wes-turner) [inline docs/model-dev-guide/profiling.rst:80]: This is nice. I think that overall, this doc is set up well.

File: docs/reference/experiment-config-reference.rst
Comment (wes-turner) [inline docs/reference/experiment-config-reference.rst:1492]: What does it do?

File: docs/.redirects/redirects.json
Comment (wes-turner) [inline docs/.redirects/redirects.json:4]: Oh -- nice thinking. I should remember this.

File: harness/determined/core/_profiler.py
Comment (wes-turner) [inline harness/determined/core/_profiler.py:166]: If this is important to you, could you please create a ticket and reference the ticket in the TODO instead of yourself? And if it's not important enough for that, then please remove the TODO.

File: harness/determined/core/_profiler.py
Comment (wes-turner) [inline harness/determined/core/_profiler.py:172]: nit: little clearer with a `defaultdict`, I think.

File: harness/determined/core/_profiler.py
Comment (wes-turner) [inline harness/determined/core/_profiler.py:246]: I'm a little surprised this works. I'd have expected that it had to look like
```python
@property
def group:
  return "network"
```

My typechecker doesn't like it, either.

File: harness/determined/core/_profiler.py
Comment (wes-turner) [inline harness/determined/core/_profiler.py:368]: nit: could you please pull this outside the try? Makes my IDE happier for the except block.

File: harness/tests/core/test_profiler.py
Comment (wes-turner) [inline harness/tests/core/test_profiler.py:148]: How do we keep this tested in CI?

File: harness/tests/core/test_profiler.py
Comment (wes-turner) [inline harness/tests/core/test_profiler.py:216]: Each test, one behavior.

File: docs/model-dev-guide/profiling.rst
Comment (azhou-determined) [reply docs/model-dev-guide/profiling.rst:75]: this can't really be configured. i left this warning from the previous doc because it's still relevant, but it has to do with being aware that there could be other experiments using the same agent as your experiment.

File: harness/tests/core/test_profiler.py
Comment (azhou-determined) [reply harness/tests/core/test_profiler.py:148]: this is tested by `test-unit-harness-gpu`

File: harness/tests/core/test_profiler.py
Comment (azhou-determined) [reply harness/tests/core/test_profiler.py:216]: sure done

File: docs/reference/experiment-config-reference.rst
Comment (wes-turner) [reply docs/reference/experiment-config-reference.rst]: Not being purposefully dense, I can't differentiate between:
```yaml
profiling:
  profiler: <val>
  enabled: <val>
```

File: docs/model-dev-guide/profiling.rst
Comment (wes-turner) [reply docs/model-dev-guide/profiling.rst:73]: I think it would be frustrating to me to read this note and then be unable to figure out how to do the thing it recommends. Let's chat about it.

File: docs/model-dev-guide/profiling.rst
Comment (wes-turner) [reply docs/model-dev-guide/profiling.rst:35]: Mmm. I missed this part.

> The Determined profiler collects a set of system metrics throughout an experiment which can be visualized in the WebUI under the experiment’s “Profiler” tab. It is supported for all training APIs, but is not enabled by default.

My fault. Like usual, I'd like someone else to blame when a thing was my fault. In that respect, could / should it have been clearer?

File: harness/determined/core/_profiler.py
Comment (wes-turner) [reply harness/determined/core/_profiler.py]: I'm a little unsettled by this. I don't think the implemented code works, either, and that it hasn't resulted in failed tests makes me wonder if something is wrong with tests.

currently implemented:
```python
    def group(self) -> str:
        return "network"
```

File: harness/determined/core/_profiler.py
Comment (wes-turner) [reply harness/determined/core/_profiler.py:172]:
```python
aggregated_metrics = defaultdict(int)
aggregated_metrics = defaultdict(defaultdict(int))
```

File: e2e_tests/tests/experiment/test_profiling.py
Comment (wes-turner) [inline e2e_tests/tests/experiment/test_profiling.py:18]: nit: I don't understand the parameterization. I think it's cleaner without it

File: docs/model-dev-guide/profiling.rst
Comment (azhou-determined) [reply docs/model-dev-guide/profiling.rst]: removed this recommendation

File: docs/reference/experiment-config-reference.rst
Comment (azhou-determined) [reply docs/reference/experiment-config-reference.rst]: removed this section

File: harness/determined/core/_profiler.py
Comment (azhou-determined) [reply harness/determined/core/_profiler.py:166]: i've created the ticket https://hpe-aiatscale.atlassian.net/browse/MD-338. i will add it to the comment but don't want to commit right now because i'm waiting for a longrunning CI to finish.

File: harness/determined/core/_profiler.py
Comment (azhou-determined) [reply harness/determined/core/_profiler.py]: fixed

File: harness/determined/core/_profiler.py
Comment (azhou-determined) [reply harness/determined/core/_profiler.py:172]: eh, don't love the way this reads. `.get(k, {})` is easily readable IMO. i do think `defaultdict` is best-practices way of doing this, but since i have a ticket to refactor this method anyway, i don't think it's worth it.

File: e2e_tests/tests/experiment/test_profiling.py
Comment (azhou-determined) [reply e2e_tests/tests/experiment/test_profiling.py:18]: done

**Review by wes-turner — APPROVED** [review]
LGTM

File: master/static/migrations/20240229114929_generic-metrics-nullable-batches.tx.down.sql
Comment (jgongd) [inline master/static/migrations/20240229114929_generic-metrics-nullable-batches.tx.down.sql:1]: The most recent migration file on 0.29.1 was named [20240306164812_remove-rm-name-expconf.tx.up.sql](https://github.com/determined-ai/determined/blob/0.29.1/master/static/migrations/20240306164812_remove-rm-name-expconf.tx.up.sql).

I suggest moving these files to the top to avoid any migration issues when users update. 
`/migration-move-to-top.sh my-migration-name`

File: proto/src/determined/api/v1/experiment.proto
Comment (jgongd) [inline proto/src/determined/api/v1/experiment.proto:635]: https://github.com/determined-ai/determined/blob/main/master/pkg/model/metrics.go#L41
https://github.com/determined-ai/determined/blob/main/master/pkg/protoutils/protoconverter/converter.go#L169

These functions may need to be updated to work with the new type.

Comment (azhou-determined) [reply proto/src/determined/api/v1/experiment.proto:635]: the READ part of getting profiling metrics from generic metrics isn't implemented in this feature, only the write. so i think it's appropriate for the time being that those functions return "unspecified" or "unknown", because at this point in time, profiling metrics should not be known to those APIs.

File: master/static/migrations/20240229114929_generic-metrics-nullable-batches.tx.down.sql
Comment (azhou-determined) [reply master/static/migrations/20240229114929_generic-metrics-nullable-batches.tx.down.sql:1]: done!

**Review by jgongd — APPROVED** [review]

**Pull request merged**
